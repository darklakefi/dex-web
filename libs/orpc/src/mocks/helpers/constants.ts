export const FIXED_SEED = "1234567890123456";
