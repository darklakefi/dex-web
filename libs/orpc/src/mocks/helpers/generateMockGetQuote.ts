import {
  randBoolean,
  randCurrencySymbol,
  randNumber,
  randUuid,
  seed,
} from "@ngneat/falso";
import type {
  GetQuoteInput,
  GetQuoteOutput,
} from "../../schemas/swaps/getQuote.schema";

export function generateMockGetQuote(input: GetQuoteInput) {
  seed(
    `${input.amountIn.toString()}${input.isXtoY.toString()}${input.slippage.toString()}${input.tokenX}${input.tokenY}${input.poolAddress}`,
  );

  return {
    amountOut: randNumber({ max: 1000000, min: 1 }),
    deadline: randNumber({ max: 1000000, min: 1 }),
    estimatedFee: randNumber({ max: 1000000, min: 1 }),
    estimatedFeesUsd: randNumber({ max: 1000000, min: 1 }),
    isXtoY: randBoolean(),
    poolAddress: randUuid(),
    priceImpactPercentage: randNumber({ max: 100, min: 1 }),
    rateXtoY: randNumber({ max: 1000000, min: 1 }),
    slippage: randNumber({ max: 100, min: 1 }),
    tokenX: {
      address: randUuid(),
      symbol: randCurrencySymbol(),
      value: randNumber({ max: 1000000, min: 1 }).toString(),
    },
    tokenY: {
      address: randUuid(),
      symbol: randCurrencySymbol(),
      value: randNumber({ max: 1000000, min: 1 }).toString(),
    },
    userAddress: randUuid(),
  } satisfies GetQuoteOutput;
}
