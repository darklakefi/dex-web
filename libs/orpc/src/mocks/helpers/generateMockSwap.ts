import {
  randBetweenDate,
  randBoolean,
  randCurrencySymbol,
  randNumber,
  randUuid,
  seed,
} from "@ngneat/falso";
import type { Swap } from "../../schemas";

export function generateMockSwap(randomSeed: string) {
  seed(randomSeed);

  return {
    buyAmount: randNumber({ max: 1000000, min: 1 }),
    buyBalance: randNumber({ max: 1000000, min: 1 }),
    buyToken: {
      address: randUuid(),
      decimals: randNumber({ max: 1000000, min: 1 }),
      symbol: randCurrencySymbol(),
    },
    estimatedFeesUsd: randNumber({ max: 1000000, min: 1 }),
    exchangeRate: randNumber({ max: 1000000, min: 1 }),
    mevProtectionEnabled: randBoolean(),
    priceImpactPercentage: randNumber({ max: 100, min: 1 }),
    sellAmount: randNumber({ max: 1000000, min: 1 }),
    sellBalance: randNumber({ max: 1000000, min: 1 }),
    sellToken: {
      address: randUuid(),
      decimals: randNumber({ max: 1000000, min: 1 }),
      symbol: randCurrencySymbol(),
    },
    slippageTolerancePercentage: randNumber({ max: 100, min: 1 }),
    swapProgressStep: randNumber({ max: 3, min: 1 }),
    swapStatus: "pending",
    swapType: "swap",
    transactionSignature: randUuid(),
    updatedAt: randBetweenDate({
      from: new Date("2025-01-01"),
      to: new Date("2025-06-24"),
    }).toISOString(),
    userAddress: randUuid(),
  } satisfies Swap;
}
