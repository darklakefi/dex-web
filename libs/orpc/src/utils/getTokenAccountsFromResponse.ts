import type { DAS } from "helius-sdk";
import { isDefined } from "remeda";

export function getTokenAccountsFromResponse(
  response: DAS.GetTokenAccountsResponse,
) {
  return response.token_accounts
    ?.map((item) => ({
      address: item.address ?? "",
      amount: item.amount ?? 0,
      mint: item.mint ?? "",
    }))
    .filter((item) => isDefined(item.amount) && isDefined(item.mint))
    .filter(Boolean);
}
