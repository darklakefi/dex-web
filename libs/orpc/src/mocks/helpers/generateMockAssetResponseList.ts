import {
  randProductDescription,
  randProductName,
  randUuid,
  seed,
} from "@ngneat/falso";
import { type DAS, Interface, OwnershipModel } from "helius-sdk";
import { times } from "remeda";
import { FIXED_SEED } from "./constants";

export function generateAssetResponseList({
  count = 10,
}: {
  count?: number;
} = {}): DAS.GetAssetResponseList {
  seed(FIXED_SEED);
  const owner = randUuid();
  return {
    cursor: "123",
    items: times(count, (_i) => ({
      burnt: false,
      content: {
        $schema: "https://schema.dexscreener.com/asset.schema.json",
        json_uri: "https://schema.dexscreener.com/asset.schema.json",
        metadata: {
          description: randProductDescription(),
          name: randProductName(),
          symbol: randProductName().slice(0, 3).toUpperCase(),
        },
      },
      id: randUuid(),
      interface: Interface.V1NFT,
      mutable: true,
      ownership: {
        delegate: undefined,
        delegated: false,
        frozen: false,
        owner,
        ownership_model: OwnershipModel.SINGLE,
      },
    })),
    limit: 10,
    total: 10,
  };
}
