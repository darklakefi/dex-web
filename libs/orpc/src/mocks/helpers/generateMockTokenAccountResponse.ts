import { randNumber, randUuid, seed } from "@ngneat/falso";
import { type DAS, Interface, OwnershipModel } from "helius-sdk";
import { times } from "remeda";
import { FIXED_SEED } from "./constants";

export function generateTokenAccountsResponseList({
  count = 10,
}: {
  count?: number;
} = {}): DAS.GetTokenAccountsResponse {
  seed(FIXED_SEED);
  return {
    limit: 10,
    token_accounts: times(count, (_i) => ({
      address: randUuid(),
      amount: randNumber({ max: 1000000, min: 1 }),
      id: randUuid(),
      interface: Interface.V1NFT,
      mint: randUuid(),
      mutable: true,
      ownership: {
        delegate: undefined,
        delegated: false,
        frozen: false,
        owner: randUuid(),
        ownership_model: OwnershipModel.SINGLE,
      },
      token_extensions: [],
    })),
    total: 10,
  };
}
