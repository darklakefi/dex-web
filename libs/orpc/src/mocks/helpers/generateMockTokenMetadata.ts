import { randCurrencySymbol, randNumber, randUuid, seed } from "@ngneat/falso";
import {
  Interface as AssetInterface,
  OwnershipModel,
} from "helius-sdk/types/enums";
import type { GetAssetResponse } from "helius-sdk/types/types";
import { FIXED_SEED } from "./constants";

export function generateMockTokenMetadata(): GetAssetResponse {
  seed(FIXED_SEED);
  return {
    burnt: false,
    id: randUuid(),
    interface: AssetInterface.V1_NFT,
    mutable: true,
    ownership: {
      delegate: undefined,
      delegated: false,
      frozen: true,
      owner: randUuid(),
      ownership_model: OwnershipModel.SINGLE,
    },
    token_info: {
      decimals: randNumber({ max: 9, min: 0 }),
      symbol: randCurrencySymbol(),
    },
  };
}
