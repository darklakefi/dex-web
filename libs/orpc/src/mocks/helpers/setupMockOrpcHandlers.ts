"use server";

import { generateMockSwap } from "./generateMockSwap";

export async function setupMockOrpcHandlers(
  page: import("@playwright/test").Page,
) {
  await page.route("**/rpc", async (route, request) => {
    const body = JSON.parse(request.postData() || "{}");

    if (body.method === "getSwapDetails") {
      const mockSwap = generateMockSwap(new Date().toISOString());

      await route.fulfill({
        body: JSON.stringify({ result: mockSwap }),
        contentType: "application/json",
        status: 200,
      });
    } else if (body.method === "getTokenDetails") {
      await route.fulfill({
        body: JSON.stringify({
          result: {
            address: "fake-address",
            imageUrl: "https://example.com/fake-token.png",
            name: "Fake Token",
            symbol: "FAKE",
            value: "1000",
          },
        }),
        contentType: "application/json",
        status: 200,
      });
    } else if (body.method === "getTokens") {
      await route.fulfill({
        body: JSON.stringify({
          result: {
            hasMore: false,
            tokens: [
              {
                address: "fake-address-1",
                imageUrl: "https://example.com/fake-token1.png",
                name: "Fake Token 1",
                symbol: "FAKE1",
                value: "1000",
              },
              {
                address: "fake-address-2",
                imageUrl: "https://example.com/fake-token2.png",
                name: "Fake Token 2",
                symbol: "FAKE2",
                value: "2000",
              },
            ],
            total: 2,
          },
        }),
        contentType: "application/json",
        status: 200,
      });
    } else if (body.method === "helius.searchAssets") {
      await route.fulfill({
        body: JSON.stringify({
          result: [
            {
              description: "Fake asset description",
              id: "asset-1",
              image: { url: "https://example.com/asset1.png" },
              name: "Fake Asset 1",
              symbol: "FAKE1",
            },
            {
              description: "Fake asset description 2",
              id: "asset-2",
              image: { url: "https://example.com/asset2.png" },
              name: "Fake Asset 2",
              symbol: "FAKE2",
            },
          ],
        }),
        contentType: "application/json",
        status: 200,
      });
    } else if (body.method === "helius.getTokenBalance") {
      await route.fulfill({
        body: JSON.stringify({
          result: {
            assets: [
              {
                description: "Fake asset description",
                id: "asset-1",
                image: { url: "https://example.com/asset1.png" },
                name: "Fake Asset 1",
                symbol: "FAKE1",
              },
            ],
            ownerAddress: "fake-owner-address",
            tokenAccounts: [
              {
                address: "account-1",
                amount: 1000,
                mint: "mint-1",
              },
            ],
            total: 1,
          },
        }),
        contentType: "application/json",
        status: 200,
      });
    } else {
      route.continue();
    }
  });
}
