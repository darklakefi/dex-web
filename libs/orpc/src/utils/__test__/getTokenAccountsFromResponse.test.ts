import { generateTokenAccountsResponseList } from "../../mocks/helpers/generateMockTokenAccountResponse";
import { getTokenAccountsFromResponse } from "../getTokenAccountsFromResponse";

const mockTokenAccountsFromResponse = generateTokenAccountsResponseList();

describe("getTokenAccountsFromResponse", () => {
  it("should return a list of assets", () => {
    const response = getTokenAccountsFromResponse(
      mockTokenAccountsFromResponse,
    );

    expect(response).toMatchSnapshot();
    expect(response?.length).toBe(
      mockTokenAccountsFromResponse.token_accounts?.length,
    );
  });
});
