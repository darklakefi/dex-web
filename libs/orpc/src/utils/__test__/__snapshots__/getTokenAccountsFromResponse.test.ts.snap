// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTokenAccountsFromResponse > should return a list of assets 1`] = `
[
  {
    "address": "3c4dfb0c-f84d-41ea-ac6d-be433685d544",
    "amount": 906466,
    "mint": "17f4a3c0-badc-4a17-8f21-89174526ced4",
  },
  {
    "address": "4cb6932d-b540-4464-b66e-b6964d4e74c9",
    "amount": 290423,
    "mint": "4bd821d2-a605-4715-b3ca-0f1a2f725144",
  },
  {
    "address": "f688dd29-19c4-4be5-b1e9-096ec3e5963c",
    "amount": 498798,
    "mint": "32821048-e405-4c14-a468-15a5dd445f9f",
  },
  {
    "address": "c78e650c-793c-43f9-ad3e-7e59d645c6c2",
    "amount": 741975,
    "mint": "352c0bb8-ab6d-4f46-8dd8-554d99d4a1a1",
  },
  {
    "address": "653cfe42-e62a-4d68-8c64-f507e3384624",
    "amount": 599623,
    "mint": "0a4ebaa4-97a4-4fc8-a90e-70ae24c17ece",
  },
  {
    "address": "96092c68-cb6d-40f2-9cbe-e2b393adfc17",
    "amount": 418322,
    "mint": "5b85dafd-df63-4696-8b02-5ed0ca7bbed7",
  },
  {
    "address": "0b110e8f-87b9-4990-8de4-0ec34c74a091",
    "amount": 155566,
    "mint": "b4622e48-1afa-44ae-be94-d19be7b39336",
  },
  {
    "address": "0e82c196-ce91-4fb4-86dd-be60ac6bce66",
    "amount": 525454,
    "mint": "85645e43-bd3f-49df-885d-8eee9e89d595",
  },
  {
    "address": "71ff1785-8ddb-4314-bb89-f4ecf04cf20e",
    "amount": 612304,
    "mint": "9bac24ee-d28d-4fd1-9727-ff524106e523",
  },
  {
    "address": "c82d3f3c-4a40-45ac-8da9-76c09db29be3",
    "amount": 4038,
    "mint": "ac1cd6bc-f3e1-4ab7-82ba-2d25cf63e63d",
  },
]
`;
